### PR TITLE
Shitcode for vpn detection 

### PR DESCRIPTION
**Pull request by DebugOk**
Something something bad code. I'm only making this a pr so I can hopefully get reviews as I finish it

TODO:
- [ ] Cache checks so the API isn't spammed

### CODE REVIEW

**Comment by DebugOk** [comment]
@DEATHB4DEFEAT if you have time, can you please take a quick look at this? I'm still working on it

**Comment by DebugOk** [comment]
I'm going to rewrite the client thing at some point but for now I just used example code :clueless:

**Comment by OCOtheOmega** [comment]
I'm very interested in this pr, so can I ask that you add the ability to whitelist players who don't need to be checked? Not IP addresses, but the players themselves.

**Comment by DebugOk** [comment]
> I'm very interested in this pr, so can I ask that you add the ability to whitelist players who don't need to be checked? Not IP addresses, but the players themselves.

I modified part of the code so that the bans it applies are marked as datacenter bans. There is a command along the lines of ban_exempt_update that you can use to apply an exclusion to a player. I just need to make it double check during the initial connect so it doesn't deny

**Comment by DEATHB4DEFEAT** [comment]
This would be great to have as a thing upstream ~~for them to maintain for us~~.
There are no downstream change markers anywhere, do you want them?

File: Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiErrorCode.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    public const int FileIoError = 68;
+
+    /// <summary>InvalidJsonResponse</summary>
+    public const int InvalidJsonResponse = 69;
+
+    /// <summary>NoData</summary>
+    public const int NoData = 70;
+
+    /// <summary>ApiGatewayError</summary>
+    public const int ApiGatewayError = 71;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiErrorCode.cs:126]: I feel like there's a better way to store this

File: Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiResponse.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        /// </summary>
+        /// <param name="errorCode"></param>
+        /// <param name="errorCause"></param>
+        /// <returns></returns>
+        public static ApiResponse OfErrorCause(int errorCode, Exception? errorCause)
+        {
+            var errorMessage = ApiErrorCode.GetErrorMessage(errorCode);
+            return new ApiResponse(default, NoContentType, NoStatus, errorCode, errorMessage, errorCause);
+        }
+    }
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiResponse.cs:113]: This is indented a tab too much

File: Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiResponse.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        /// Create an API response for error cause
+        /// </summary>
+        /// <param name="errorCode"></param>
+        /// <param name="errorCause"></param>
+        /// <returns></returns>
+        public static ApiResponse OfErrorCause(int errorCode, Exception? errorCause)
+        {
+            var errorMessage = ApiErrorCode.GetErrorMessage(errorCode);
+            return new ApiResponse(default, NoContentType, NoStatus, errorCode, errorMessage, errorCause);
+        }
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/NeutrinoApi/ApiResponse.cs:112]: Blank xmldoc arg comments

File: Content.Server/DeltaV/ProxyDetection/DetectionManager.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    [Dependency] private readonly ILocalizationManager _loc = default!;
+    [Dependency] private readonly ILogManager _log = default!;
+    [Dependency] private readonly IBanManager _banManager = default!;
+
+    private ISawmill _sawmill = default!;
+
+    private bool _shouldProbe;
+    private bool _apiValid;
+
+    private NeutrinoApiClient _neutrinoApiClient = default!;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/DetectionManager.cs:29]:
```suggestion
    private ISawmill _sawmill = default!;
    private NeutrinoApiClient _neutrinoApiClient = default!;

    private bool _shouldProbe;
    private bool _apiValid;
```

File: Content.Server/DeltaV/ProxyDetection/DetectionManager.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    void IPostInjectInit.PostInject()
+    {
+        _sawmill = _log.GetSawmill("PROXYDETECTION");
+
+        if (_cfg.GetCVar(DCCVars.BlockProxyConnections))
+        {
+            if (_cfg.GetCVar(DCCVars.ProxyApiKey) == "" || _cfg.GetCVar(DCCVars.ProxyApiUser) == "")
+                return;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/DetectionManager.cs:119]: This exactly probably won't work, but these ifs can be compacted to one
```suggestion
        if (_cfg.GetCVar(DCCVars.BlockProxyConnections) &&
            !(_cfg.GetCVar(DCCVars.ProxyApiKey) == "" || _cfg.GetCVar(DCCVars.ProxyApiUser) == ""))
```

File: Content.Server/DeltaV/ProxyDetection/NeutrinoApi/NeutrinoApiClient.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    /// </summary>
+    /// <param name="httpMethod"></param>
+    /// <param name="endpoint"></param>
+    /// <param name="paramDict"></param>
+    /// <param name="outputFilePath"></param>
+    /// <param name="timeoutInSeconds"></param>
+    /// <returns>ApiResponse object on success or failure</returns>
+    private async Task<ApiResponse> ExecRequest(string httpMethod, string endpoint,
+        Dictionary<string, string> paramDict, int timeoutInSeconds)
+    {
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/NeutrinoApi/NeutrinoApiClient.cs:84]: Blank xmldoc arg comments

File: Content.Server/DeltaV/ProxyDetection/NeutrinoApi/NeutrinoApiClient.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            using var request = new HttpRequestMessage();
+            request.Method = new HttpMethod(httpMethod.Equals("POST") ? "POST" : "GET");
+            request.RequestUri = new Uri(url);
+            request.Content = httpMethod.Equals("POST") ? encodedParams : null;
+            using var cancellationToken = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutInSeconds));
+            using var response = await Client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead,
+                cancellationToken.Token);
+            var content = await response.Content.ReadAsStringAsync(cancellationToken.Token);
+            var statusCode = (int) response.StatusCode;
+            var contentType = response.Content.Headers.ContentType?.ToString();
```
Comment (DEATHB4DEFEAT) [inline Content.Server/DeltaV/ProxyDetection/NeutrinoApi/NeutrinoApiClient.cs:106]:
```suggestion
            using var request = new HttpRequestMessage();
            request.Method = new HttpMethod(httpMethod.Equals("POST") ? "POST" : "GET");
            request.RequestUri = new Uri(url);
            request.Content = httpMethod.Equals("POST") ? encodedParams : null;

            using var cancellationToken = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutInSeconds));
            using var response = await Client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead,
                cancellationToken.Token);

            var content = await response.Content.ReadAsStringAsync(cancellationToken.Token);
            var statusCode = (int) response.StatusCode;
            var contentType = response.Content.Headers.ContentType?.ToString();
```

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

**Pull request closed without merging**